[clang-cpp] Assertion failure in adjust_cpp_member when verifying ifstream_get_line_1 reproducer

## Summary

ESBMC crashes with an internal assertion failure when verifying the
`regression/esbmc-cpp/stream/ifstream_get_line_1/` reproducer:

```
Assertion failed: (comp_symb), function adjust_cpp_member,
                  file clang_cpp_adjust_expr.cpp, line 198.
```

The assertion fires from `clang_cpp_adjust::adjust_cpp_member` when
`ns.lookup(expr.component_name())` returns null — i.e. the C++
frontend is trying to lower a `member_exprt` for a method call whose
component symbol it cannot resolve.

## Reproducer

The sources are already in the repository:

```sh
cd regression/esbmc-cpp/stream/ifstream_get_line_1
esbmc lista7paa.cpp Solver.cpp Parser.cpp \
  --unwind 1 --no-pointer-check --no-bounds-check \
  --no-unwinding-assertions --timeout 60
```

Output:

```
Parsing lista7paa.cpp
Parsing Solver.cpp
Parsing Parser.cpp
Parser.cpp:202:11: warning: ISO C++11 does not allow conversion from string literal to 'char *'
Parser.cpp:202:26: warning: ISO C++11 does not allow conversion from string literal to 'char *'
Converting
Assertion failed: (comp_symb), function adjust_cpp_member, file clang_cpp_adjust_expr.cpp, line 198.
```

## Context

Discovered while sweeping the C++ `KNOWNBUG` regressions tracked in
#4401. The `ifstream_get_line_1` test's `test.desc` was passing
`Solver.cpp Parser.cpp` only — omitting `lista7paa.cpp` (the file
containing `main`) — so on master the test fails harmlessly with
`ERROR: main symbol 'main' not found`. Adding the missing source
file exposes this deeper crash, which is unrelated to the
stream/string operational models being addressed in #4401, so it is
filed here as a standalone frontend bug.

## What `adjust_cpp_member` does

The relevant code:

```cpp
void clang_cpp_adjust::adjust_cpp_member(member_exprt &expr)
{
  // ... replaces OBJECT.setX() / OBJECT->setX() with the symbol expr
  // for setX. Requires the component to resolve.
  const symbolt *comp_symb = ns.lookup(expr.component_name());
  assert(comp_symb);
  ...
}
```

It is reached during conversion of one of the three .cpp files
(likely `Parser.cpp`, which is the one whose warnings appear
immediately before the crash). The component symbol is presumably
the clang ID of a method that the frontend has not registered in
the symbol table by the time `adjust_expr` runs.

## Suggested first steps

1. Replace the `assert` with a structured frontend error so the
   offending source line and component name are reported instead of
   crashing the binary.
2. Find which member call in `Parser.cpp` (or transitive includes)
   produces the unresolved component_name — likely a templated
   `std::string` or stream method that the C++ operational model
   does not register at the expected ID.

## Once fixed

Re-add `lista7paa.cpp` to the test.desc source list, drop the
`KNOWNBUG` line, tick the box in #4401.

--- a/regression/esbmc-cpp/vector/github_4416/holder.cpp
+++ b/regression/esbmc-cpp/vector/github_4416/holder.cpp
@@ -1,0 +1,12 @@
+#include "holder.h"
+
+// `tab.at(i).resize(inner)` is the first instantiation of
+// `std::vector<bool>::resize(int)` in this build. The crash in
+// adjust_cpp_member fired here when `main.cpp` was parsed first and
+// completed `vector<bool>` without seeing this method.
+holder::holder(int outer, int inner)
+{
+  tab.resize(outer);
+  for (int i = 0; i < outer; ++i)
+    tab.at(i).resize(inner);
+}

--- a/regression/esbmc-cpp/vector/github_4416/holder.h
+++ b/regression/esbmc-cpp/vector/github_4416/holder.h
@@ -1,0 +1,12 @@
+#ifndef GITHUB_4416_HOLDER_H_
+#define GITHUB_4416_HOLDER_H_
+
+#include <vector>
+
+class holder {
+public:
+  holder(int outer, int inner);
+  std::vector<std::vector<bool> > tab;
+};
+
+#endif

--- a/regression/esbmc-cpp/vector/github_4416/main.cpp
+++ b/regression/esbmc-cpp/vector/github_4416/main.cpp
@@ -1,0 +1,7 @@
+#include "holder.h"
+
+int main()
+{
+  holder h(2, 3);
+  return 0;
+}

--- a/regression/esbmc-cpp/vector/github_4416/test.desc
+++ b/regression/esbmc-cpp/vector/github_4416/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+holder.cpp --unwind 1 --no-pointer-check --no-bounds-check --no-unwinding-assertions
+^VERIFICATION SUCCESSFUL$

--- a/src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp
@@ -4,6 +4,7 @@
 #include <util/c_types.h>
 #include <util/destructor.h>
 #include <util/expr_util.h>
+#include <util/message.h>
 
 clang_cpp_adjust::clang_cpp_adjust(contextt &_context)
   : clang_c_adjust(_context)
@@ -195,7 +196,18 @@ void clang_cpp_adjust::adjust_cpp_member(member_exprt &expr)
    *      * id: <setX_clang_ID>
    */
   const symbolt *comp_symb = ns.lookup(expr.component_name());
-  assert(comp_symb);
+  if (!comp_symb)
+  {
+    log_error(
+      "{}: unresolved C++ member component `{}` at {} (struct-op type "
+      "id `{}`, member type id `{}`)",
+      __func__,
+      expr.component_name(),
+      expr.location().as_string(),
+      expr.struct_op().type().id_string(),
+      expr.type().id_string());
+    abort();
+  }
   // compoment's type shall be the same as member_exprt's type
   // and both are of the type `code`
   assert(comp_symb->type.is_code());

--- a/src/clang-cpp-frontend/clang_cpp_convert.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_convert.cpp
@@ -2225,6 +2225,44 @@ bool clang_cpp_convertert::get_member_expr(
   if (perform_virtual_dispatch(memb))
     return get_vft_binding_expr(memb, new_expr);
 
+  // Lazy-register a body-less symbol for member calls whose enclosing
+  // class was already completed by an earlier translation unit. Without
+  // this, a template method first instantiated in *this* TU (e.g.
+  // `std::vector<bool>::resize` first used in Solver.cpp after the class
+  // was completed while parsing lista7paa.cpp) would never be added to
+  // the context, and `adjust_cpp_member` would later fail to resolve it
+  // (issue #4416). The signature is enough: the body-less stub is sound
+  // because symex falls back to nondet for the return and havocs pointer
+  // args at call sites with `body_available=false`. Constructors
+  // are unreached here (CXXConstructExpr path); destructors and lambda
+  // `operator()` go through this branch and are handled identically.
+  // See also `get_method` for the canonical full-body version.
+  if (
+    const auto *md = llvm::dyn_cast<clang::CXXMethodDecl>(memb.getMemberDecl()))
+  {
+    std::string id, name;
+    get_decl_name(*md, name, id);
+    if (!id.empty() && context.find_symbol(id) == nullptr)
+    {
+      typet sym_type;
+      if (get_type(md->getType(), sym_type))
+        return true;
+      symbolt symbol;
+      locationt loc;
+      get_location_from_decl(*md, loc);
+      get_default_symbol(
+        symbol,
+        get_modulename_from_path(loc.file().as_string()),
+        sym_type,
+        name,
+        id,
+        loc);
+      symbol.lvalue = true;
+      symbol.is_extern = true;
+      context.move_symbol_to_context(symbol);
+    }
+  }
+
   return clang_c_convertert::get_member_expr(memb, new_expr);
 }
 

